Start QEMU emulation in EL3

carlocaione opened the pull request:
QEMU support for EL3 is mature enough that we can start directly from EL3 and drop to EL1 in the reset routine. Remove also the legacy SWITCH_TO_EL1 because there is no need to have Zephyr running at EL3.

This makes the emulation also closer to what happens on real hardware where the CPU starts usually from EL3.

carlocaione committed:
boards: qemu_cortex_a53: Start in EL3

No need anymore to start QEMU in EL1. Start directly from EL3 and let
the code drop to EL1 like it would happen on a real hardware.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arch: arm64: Remove CONFIG_SWITCH_TO_EL1

Remove the useless CONFIG_SWITCH_TO_EL1 since there should be no reason
to run Zephyr in EL3. So just drop to EL1 by default when booting from
EL3. Remove also non-reachable code.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

sandeepbrcm reviewed (commented):
So are you also going to remove rest of the el3 code as they become stale after this.

carlocaione commented:
> So are you also going to remove rest of the el3 code as they become stale after this.

Yeah, good point. I think we can remove all the EL2 code as well since we are dropping in EL1 anyway at start.

carlocaione committed:
arch: arm64: Remove EL2/EL3 code

Zephyr is only supposed to be running at EL1 (+ EL0). Now that we drop
in EL1 from ELn at start we can remove all the EL2/EL3 unused code.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

Abhishek-brcm commented on arch/arm/core/aarch64/fatal.c:
```diff
@@ -162,36 +162,17 @@ void z_arm64_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
-		switch (GET_EL(el)) {
-		case MODE_EL1:
+		if (GET_EL(el) != MODE_EL0) {
 			__asm__ volatile("mrs %0, esr_el1" : "=r" (esr));
```
would recommend using `read_sysreg`/`write_sysreg` APIs for consistency, maybe a separate patch.

carlocaione replied:
Ok, I'll add a new patch on top.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -119,7 +118,7 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+
 	/*
 	 * Enable the instruction cache, stack pointer and data access
 	 * alignment checks.
```
EL2 code from `reset.S` also should be removed?

carlocaione replied:
In theory you could start QEMU in EL2 and drop to EL1 from there. So I'd leave that in place.

carlocaione replied:
Uhm, actually I just realized that from EL2 we are not dropping to EL1. At this point I'd just deprecate EL2 and remove also that code path.

carlocaione committed:
arch: arm64: Deprecate booting from EL2

We are deprecating booting and running in EL2.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

Abhishek-brcm commented on arch/arm/core/aarch64/fatal.c:
```diff
@@ -163,12 +163,13 @@ void z_arm64_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
-			__asm__ volatile("mrs %0, far_el1" : "=r" (far));
-			__asm__ volatile("mrs %0, elr_el1" : "=r" (elr));
-
+			esr = read_sysreg(esr_el1);
```
I meant replacing __asm__ volatile everywhere in arm64 code with read_sysreg/write_sysreg, not just this file. Can take it up later as improvement if not now.

carlocaione replied:
hooo ok. I'm preparing a PR just for cleanup stuff, I'll add the commit in that PR. For this PR I think this is enough.

Abhishek-brcm replied:
In that case, I recommend removing this commit from current PR and taking care of it in a single commit in next PR.

carlocaione replied:
ACK. Removing it.

Abhishek-brcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -119,39 +118,10 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
-	msr	sctlr_el2, x0
-	b	0f
+	/* Booting from EL2 is not supported */
+	b	.
```
I am bit skeptical about this (b .) as user will not have any idea where system hung. But I do not have alternate idea at this point. (there may be some idea around the fact that we should be able to access EL1 registers from EL2 if accessing and assert at later point where we are able to print the fault)

Everything else looks fine.

sandeepbrcm reviewed (approved):
looks good in scope of this pr

carlocaione commented:
gentle ping about this PR
